### PR TITLE
[tab-bar] fix widget leaking via phosphor VDOM

### LINKED ISSUE #5129: widgets seem to leak via tab bar toolbar contribution
Particularly Monaco editors:
Screen Shot 2019-05-13 at 15 01 02

### PR DESCRIPTION
**Pull request by akosyakov**
fix #5129

In order to verify one can open several editors and close them, do it several times, then collect the garbage and check the memory snapshot how many instances of `MonacoEditor` are there.

### CODE REVIEW

File: packages/core/src/browser/shell/tab-bars.ts [hunk 1]
```diff
@@ -86,8 +86,8 @@ export class TabBarRenderer extends TabBar.Renderer {
         const dataset = this.createTabDataset(data);
         return h.li(
             {
-                key, className, title: title.caption, style, dataset,
-                oncontextmenu: event => this.handleContextMenuEvent(event, title)
+                key, className, id: title.owner.id, title: title.caption, style, dataset,
```
Comment (akosyakov) [inline packages/core/src/browser/shell/tab-bars.ts]: that's a bad idea, we need to add some prefix, otherwise there are more than one node with widget id

**Comment by kittaakos** [comment]
I am checking this...

**Comment by kittaakos** [comment]
Looks good so far. Perhaps you can reduce the visibility of the `createTabId` method. It is not called outside of the class. I did not see `StandaloneCodeEditor` instances, but after a refresh, a GC collection and a heap dump, I can see two diff editor instances. Do you want to check or it's unrelated to this PR?

Screen Shot 2019-05-13 at 17 40 05

----

Question, do we have to mimic from now on the pattern you did for the dock panel, the veriable resolver, and the tab-bar toolbar widget? Namely this
```ts
this.toDisposeOnMarkAsCurrent.dispose();
this._currentTitle = title;
if (title) {
    const resetCurrent = () => this.markAsCurrent(undefined);
    title.owner.disposed.connect(resetCurrent);
    this.toDisposeOnMarkAsCurrent.push(Disposable.create(() =>
        title.owner.disposed.disconnect(resetCurrent)
    ));
}
```
When do we have to do the same in the future to avoid ponntentional memory leaks? Just guessing, we cannot just save a widget as a reference anymore. Thanks!

**Comment by AlexTugarev** [comment]
> but after a refresh ...

@kittaakos, do you mean browser refresh? if so, maybe related to leaking dev tools?

**Comment by AlexTugarev** [comment]
@akosyakov, I've checked as well. No retainers from tab-bar. 

But there still are some left-overs connected to editor preview widget:

Screen Shot 2019-05-13 at 17 58 54

**Comment by akosyakov** [comment]
> When do we have to do the same in the future to avoid ponntentional memory leaks? Just guessing, we cannot just save a widget as a reference anymore. Thanks!

@kittaakos i don't think it is a performance issue, but rathe functional. A reference usually gets overloaded by next widget, so there is no really a memory leak. But if you close a widget then commands, keybindings, when closures and resolved values should not behave like the widget is still opened.

**Comment by akosyakov** [comment]
@AlexTugarev @kittaakos Would it be fine to tackle diff and preview editors leaking separately? They are not so invasive as VDOM leaking any widget with a toolbar.

**Comment by kittaakos** [comment]
> do you mean browser refresh? if so, maybe related to leaking dev tools?

Yeah, after #4434 I feel a little bit uncomfortable with these memory reviews :)

> i don't think it is a performance issue, but rathe functional. A reference usually gets overloaded by next widget, so there is no really a memory leak. But if you close a widget then commands, keybindings, when closures and resolved values should not behave like the widget is still opened.

Sorry for the confusion, let me rephrase. Can you give us a clear guide when to you the patter you have introduced to fix the memory problem? Something that could be referenced later.

> Would it be fine to tackle diff and preview editors leaking separately?

Absolutely. That's why asked whether you want to check in this PR or not.
Still, I think you could reduce the visibility of the method I have mentioned above, and then from my side, it's OK to merge.

**Comment by akosyakov** [comment]
Yes, you can use this pattern to clean up a widget reference from global services.

I've reduced the visibility of a new method to `protected`.
